File: question/neetcode/binary_tree/validate-binary-search-tree.py
from typing import Optional
import math

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """
                3
            2       4
        1               5

        min=-inf max=inf
        if not min<node<max return false
        else dfs(left) and dfs(right)
        2 < 3
        """

        def dfs(node: Optional[TreeNode], min_: int, max_: int):
            if not node:
                return True
            
            if not (min_<node.val<max_):
                return False
            
            return dfs(node.left,min_,node.val) and dfs(node.right,node.val,max_)
        
        return dfs(root,-math.inf,math.inf)

File: question/neetcode/binary_tree/validate-binary-search-tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool dfs(TreeNode* node, long long l, long long r){
        if(not node) return true;

        if(node->val<=l or node->val>=r) return false;

        return dfs(node->left,l,node->val) and dfs(node->right,node->val,r);
    }

    bool isValidBST(TreeNode* root) {
        return dfs(root, -LONG_LONG_MAX, LONG_LONG_MAX);
    }
};